firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
+            eq = FIAT.FDMQuadrature(ref_el, degree)
+            e0 = elements[0] if elements[0].formdegree == 0 else FIAT.FDMLagrange(ref_el, degree)
+            e1 = elements[-1] if elements[-1].formdegree == 1 else FIAT.FDMDiscontinuousLagrange(ref_el, degree-1)
+            if is_interior:
+                e0 = FIAT.RestrictedElement(e0, restriction_domain="interior")
+            if hasattr(eq.dual, "rule"):
+                rule = eq.dual.rule
+            else:
+                rule = FIAT.quadrature.make_quadrature(ref_el, degree+1)
+
+            pts = rule.get_points()
+            wts = rule.get_weights()
+
+            phiq = eq.tabulate(0, pts)
+            phi1 = e1.tabulate(0, pts)
+            phi0 = e0.tabulate(1, pts)
+
+            moments = lambda v, u: numpy.dot(numpy.multiply(v, wts), u.T)
+            A00 = moments(phiq[(0, )], phi0[(0, )])
+            A11 = moments(phi1[(0, )], phi1[(0, )])
+            A10 = moments(phi1[(0, )], phi0[(1, )])
+            A10 = numpy.linalg.solve(A11, A10)
+            A11 = numpy.eye(A11.shape[0])
+
+            Ihat = mass_matrix(tdim, formdegree, A00, A11)
+            Dhat = diff_matrix(tdim, formdegree, A00, A11, A10)
+            result = block_mat([[Ihat], [Dhat]])
+            Ihat.destroy()
+            Dhat.destroy()
+
+            if value_size != 1:
+                eye = petsc_sparse(numpy.eye(value_size))
+                temp = result
+                result = temp.kron(eye)
+                temp.destroy()
+                eye.destroy()
+
+            if is_facet:
+                cache[full_key] = result
+                noperm = PETSc.IS().createGeneral(numpy.arange(result.getSize()[0], dtype=PETSc.IntType), comm=result.comm)
+                result = result.createSubMatrix(noperm, self.ises[1])
+                noperm.destroy()
+
+            return cache.setdefault(key, result)
+
+
+def factor_interior_mat(A00):
+    # Assume that interior DOF list i0 is ordered such that A00 is block diagonal
+    # with blocks of increasing dimension
+    indptr, indices, data = A00.getValuesCSR()
+    degree = numpy.diff(indptr)
+
+    # TODO handle non-symmetric case with LU, requires scipy
+    invchol = lambda X: numpy.linalg.inv(numpy.linalg.cholesky(X))
+    nblocks = numpy.count_nonzero(degree == 1)
+    zlice = slice(0, nblocks)
+    numpy.sqrt(data[zlice], out=data[zlice])
+    numpy.reciprocal(data[zlice], out=data[zlice])
+    PETSc.Log.logFlops(2*nblocks)
+    for k in range(2, degree[-1]+1):
+        nblocks = numpy.count_nonzero(degree == k)
+        zlice = slice(zlice.stop, zlice.stop + k*nblocks)
+        data[zlice] = invchol(data[zlice].reshape((-1, k, k))).reshape((-1,))
+        flops = ((k+1)**3 + 5*(k+1)-12)//3 + k**3
+        PETSc.Log.logFlops(flops*nblocks)
+
+    A00.setValuesCSR(indptr, indices, data)
+    A00.assemble()
+
+
+@PETSc.Log.EventDecorator("FDMCondense")
+def condense_element_mat(A, i0, i1, submats):
```

Comment:
Comment "Return the Schur complement associated to indices in i1, condensing i0 out"